xarray/backends/api.py

```diff
@@ -617,8 +617,15 @@ def to_netcdf(dataset, path_or_file=None, mode='w', format=None, group=None,
     # if a writer is provided, store asynchronously
     sync = writer is None
 
+    # TODO Move this logic outside of this function
+    from .common import get_scheduler, get_scheduler_lock
+    scheduler = get_scheduler()
+    lock = get_scheduler_lock(scheduler)(path_or_file)
```

Comment:
It occurs to me that this becomes even more complicated if we allow for reading/writing netCDF3 files with SciPy, which certainly does not need an HDF5 lock. But perhaps it's not worth worrying about that yet.

Reply:
> Only a single process will be creating file_lock, so we don't need to worry about including the filename as part of the lock

Do you ever save to multiple files at the same time?  Are we locking around the file or around the HDF5 library?  If we have many files and we're locking around the file then having a different lock per file might reduce contention, especially if there are very many files.

Reply:
Yes, there is the [`save_mfdataset`](http://xarray.pydata.org/en/latest/generated/xarray.save_mfdataset.html#xarray.save_mfdataset) function. I haven't gotten there yet but this is a use case we'll want to support.

Reply:
> Are we locking around the file or around the HDF5 library?

Per process, only one thread can use the HDF5 library (it is not threadsafe). Across multiple processes, there can be multiple readers per file but only one writer.

Reply:
I see, so your choice of ordering locks here is quite intentional:

```python
lock = CombinedLock([HDF5_LOCK, file_lock])
```

Grab the lock for this process first, then grab the lock for the file that we're trying to read.

Reply:
That's a good point that the ordering of the locks may matter. I hadn't thought that through yet.

It might actually be better to lock files first, if we expect the HDF5 lock to have more contention. I suppose this could be answered experimentally.

Potentially it could be useful for the dask scheduler to understand these limitations, too, by encouraging it to schedule writes to the same file in the same process.